rfcs/20231005-hybrid-quantization.md

```diff
@@ -0,0 +1,116 @@
+quantization because of the following reasons.
+
+* It aligns well with the existing behavior of the corresponding HLO ops which
+allow operand types to differ and upcast operands to the highest-precision type
+in such a case.
+* Being able to express weight-only natively in StableHLO, it does not rely on
+pattern matching by backend to avoid patterns that might result in bad
+performance.
+* Constant and dequantize can unintentionally be folded in downstreams and
+hybrid op removes concern on constant folding. We can also consider inserting
+optimization_barrier in between to prevent constant folding, but this requires
+upstream frameworks to embed optimization_barrier and downstreams to pattern
+match additional optimization_barrier. It is hard to guarantee these
+requirements on use-cases across various frameworks and hardwares.
+
+## Examples
+
+Here are examples of hybrid quantized convolution and dot_general. Hybrid
+quantized ops will get float input(or lhs) and quantized weight(or rhs) as
+operands and output float results.
+
+```mlir
+%conv = "stablehlo.convolution"(%input_f, %weight_q) ... : (tensor<...xf32>, !quant.uniform<i8:f32, scale:zp>) -> tensor<...xf32>>
+```
+
+```mlir
+%dot = "stablehlo.dot_general"(%lhs_f, %rhs_q) ... : (tensor<...xf32>, !quant.uniform<i8:f32, scale:zp>) -> tensor<...xf32>>
+```
+
+## Proposed spec changes
+
+We propose to modify a few constraints on quantized convolution and dot_general
+to represent weight-only quantization using hybrid op.
+
+### convolution
+
+#### Current constraints
+
+* If the operation uses quantized tensors:
+  * (C28) `is_quantized_tensor(lhs) and is_quantized_tensor(rhs) and
+    is_quantized_tensor(result)`.
+  * (C29) `storage_type(lhs) =  storage_type(rhs)`.
+  * (C30) `expressed_type(lhs) = expressed_type(rhs) = expressed_type(result)`.
+  * (C31) If `is_per_tensor_quantized(rhs)`,
+    then `is_per_tensor_quantized(result)`.
+  * (C32) If `is_per_axis_quantized(rhs)`, then
+    `quantization_dimension(rhs) = kernel_output_feature_dimension`.
+  * (C33) If `is_per_axis_quantized(result)`, then
+    `quantization_dimension(result) = output_feature_dimension`.
+
+#### Proposed constraints
+
+* If the operation uses quantized tensors:
+  * (C28) `is_quantized(lhs) = is_quantized(result) and is_quantized(rhs)`.
+  * (C29) `If is_per_axis_quantized(rhs)`,
+    then `quantization_dimension(rhs) = kernel_output_feature_dimension`.
+  * (C30) `If is_per_axis_quantized(result)`, then 
+    `quantization_dimension(result) = output_feature_dimension`.
+  * If `is_quantized(lhs)`:
+    * (C31) `storage_type(lhs) = storage_type(rhs)`.
+    * (C32) `expressed_type(lhs) = expressed_type(rhs) = expressed_type(result)`.
+    * (C33) If `is_per_tensor_quantized(rhs)`, then
+      `is_per_tensor_quantized(result)`.
+  * If `!is_quantized(lhs)`:
+    * (C34) `element_type(lhs) = expressed_type(rhs) = element_type(result)`.
```

Comment:
Need to update the semantics description as well.

The current text is:

> For quantized types, performs `dequantize_op_quantize( lambda lhs, rhs: convolution(lhs, rhs, window_strides, padding, lhs_dilation, rhs_dilation, window_reversal, input_batch_dimension, input_feature_dimension, input_spatial_dimensions, kernel_input_feature_dimension, kernel_output_feature_dimension, kernel_spatial_dimensions, output_batch_dimension, output_feature_dimension, output_spatial_dimensions, feature_group_count, batch_group_count, precision_config), lhs, rhs, type(result))`.

This description (and the definition of `dequantize_op_quantize` at the end of spec) can only handle quantized operations with all quantized inputs/outputs.

Reply:
+1 the semantics needs update as well. here and below. 
As logan mentioned `dequantize_op_quantize` dequantizes all the inputs and produce a quantized output.  For quantized types, the semantics should accommodate both the weight-only and all-inputs-quantized cases.

Reply:
Agree that the semantics need to be updated as well. I added `dequantize_op` and updated semantics using it. Could you take a look?